e2e-tests/upgrade/run

```diff
@@ -8,117 +8,182 @@ test_dir=$(realpath $(dirname $0))
         sleep 5
     done
-
-    if [[ "${IMAGE_TO_UPDATE}" == $(kubectl_bin get pod --selector=name="${OPERATOR_NAME}" -o jsonpath='{.items[*].spec.containers[?(@.name == "'"${OPERATOR_NAME}"'")].image}') \
-          && "${IMAGE_MONGOD}" == $(kubectl_bin get psmdb "${CLUSTER}" -o jsonpath='{.spec.image}') ]]; then
-        : Operator image has been updated correctly
-    else
-        echo 'Operator image has not been updated'
-        exit 1
-    fi
-    
-    compare_generation 1 statefulset "${CLUSTER}-rs0"
-
     sleep 10
+
+    desc 'check images and generation after operator upgrade'
     wait_for_running "${CLUSTER}-rs0" "${CLUSTER_SIZE}"
+    check_applied_images "operator"
+    compare_generation "1" "statefulset" "${CLUSTER}-rs0"
+    compare_generation "1" "psmdb" "${CLUSTER}"
 
+    desc 'patch psmdb images and upgrade'
     kubectl_bin patch psmdb "${CLUSTER}" --type=merge --patch '{
-        "metadata": {"annotations":{ "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"'"$API_TO_UPDATE"'\"}" }},
         "spec": {
-            "image": "'"$IMAGE_MONGOD_TO_UPDATE"'",
-            "pmm": { "image": "'"$IMAGE_PMM_TO_UPDATE"'" },
-            "backup": { "image": "'"$IMAGE_BACKUP_TO_UPDATE"'" }
+            "crVersion": "'"${TARGET_OPERATOR_VER}"'",
+            "initImage": "'"${TARGET_IMAGE}"'",
+            "image": "'"${TARGET_IMAGE_MONGOD}"'",
+            "pmm": { "image": "'"${TARGET_IMAGE_PMM}"'" },
+            "backup": { "image": "'"${TARGET_IMAGE_BACKUP}"'" }
         }}'
-
-    sleep 10
-    compare_generation 2 statefulset "${CLUSTER}-rs0"
-    compare_generation 2 psmdb "${CLUSTER}"
-
-    if [[ "${IMAGE_TO_UPDATE}" == $(kubectl_bin get pod --selector=name="${OPERATOR_NAME}" -o jsonpath='{.items[*].spec.containers[?(@.name == "'"${OPERATOR_NAME}"'")].image}') \
-          && "${IMAGE_MONGOD_TO_UPDATE}" == $(kubectl_bin get psmdb "${CLUSTER}" -o jsonpath='{.spec.image}') ]]; then
-        : MongoDB image has been updated correctly
-    else
-        echo 'MongoDB image has not been updated'
-        exit 1
-    fi
-    
     sleep 10
 
-    wait_cluster_consistency "${CLUSTER}" "${CLUSTER_SIZE}"
-    sleep 10
-    wait_cluster_consistency "${CLUSTER}" "${CLUSTER_SIZE}"
-    sleep 10 
-    wait_cluster_consistency "${CLUSTER}" "${CLUSTER_SIZE}"
-
+    desc 'check cluster after full upgrade'
     wait_for_running "${CLUSTER}-rs0" "${CLUSTER_SIZE}"
-
+    wait_cluster_consistency "${CLUSTER}" "${CLUSTER_SIZE}"
     simple_data_check "${CLUSTER}-rs0" "${CLUSTER_SIZE}"
-    destroy $namespace
+    check_applied_images "all"
+    compare_generation "2" "statefulset" "${CLUSTER}-rs0"
+    compare_generation "2" "psmdb" "${CLUSTER}"
+
+    desc 'cleanup'
+    kubectl_bin delete -f "${cr_yaml}" || true
+    kubectl_bin delete -f "${src_dir}/deploy/crd.yaml" || true
+    kubectl_bin delete pvc --all
+    kubectl_bin delete -f "${src_dir}/deploy/rbac.yaml" || true
```

Comment:
Why do you need do delete all these stuff manually? Maybe it is better to just use  `destroy "${namespace}"` or improve  destroy function if it is needed.

Reply:
moved to destroy function